tests/testthat/test-renderer1-knit-print.R

```diff
@@ -1,168 +1,172 @@
+test_that("id property is unique over entire page", {
+  expect_equal(length(not.unique), 0)
+})
+
+test_that("id must contain at least one character", {
+  expect_true(all(0 < nchar(names(id.counts))))
+})
+
+test_that("id must not contain any space characters", {
+  expect_false(any(grepl(" ", names(id.counts))))
+})
+
+## function to extract all circles from an HTML page
+get_circles <- function(html=getHTML()) {
+  plot.names <- c("plot1top", "plot1bottom")
+  count.vec <- c()
+  for(i in seq_along(plot.names)){
+    xpath <- sprintf("//div[@id='%s']//circle[@class='geom']", plot.names[[i]])
+    circle.list <- getNodeSet(html, xpath)
+    count.vec[[i]] <- length(circle.list)
+  }
+  count.vec
+}
+
+get_elements <- function(id){
+  a<- runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('show_hide_selector_widgets')[0]", as.character(id)),return.value=TRUE)
+  b<- runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('show_hide_selector_widgets')[0]", as.character(id)),return.value=TRUE)
+  show_hide <- runtime_evaluate(script =sprintf("div = document.getElementById('%s');div.getElementsByClassName('table.legend tr.label_variable')[0]", as.character(id)), return.value=TRUE)
+  widget <-runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('table.legend tr.label_variable')[1]", as.character(id)),return.value=TRUE)
+  list(a178=a,
+       b934=b,
+       show_hide=show_hide,
+       widget=widget)
+}
+
+plot1top <- get_elements("plot1top")
+plot1bottom <- get_elements("plot1bottom")
+
+# get_circles() returns a list of circles so comparing with list(expected, expected)
+
+test_that("clicking top legend adds/remove points", {
+  expect_equal(get_circles(), list(10, 10))
+  
+  clickID("plot1top_q_label_variable_a178")
+  expect_equal(get_circles(), list(5, 10))
+  
+  clickID("plot1top_q_label_variable_b934")
+  expect_equal(get_circles(), list(0, 10))
+  
+  clickID("plot1top_q_label_variable_b934")
+  expect_equal(get_circles(), list(5, 10))
+  
+  clickID("plot1top_q_label_variable_a178")
+  expect_equal(get_circles(), list(10, 10))
+})
+
+test_that("clicking bottom legend adds/remove points", {
+  expect_equal(get_circles(), list(10, 10))
+  
+  clickID("plot1bottom_q_label_variable_a178")
+  expect_equal(get_circles(), list(10, 5))
+  
+  clickID("plot1bottom_q_label_variable_b934")
+  expect_equal(get_circles(), list(10, 0))
+  
+  clickID("plot1bottom_q_label_variable_b934")
+  expect_equal(get_circles(), list(10, 5))
+  
+  clickID("plot1bottom_q_label_variable_a178")
+  expect_equal(get_circles(), list(10, 10))
+})
+clickSide<- function(position=NULL){
+  if(position=="top"){
+    runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('show_hide_selector_widgets')[0].dispatchEvent(new CustomEvent('click'));", as.character("plot1top")))
```

Comment:
again repeated JS code, please refactor to use for loop or helper function

Reply:
> again repeated JS code, please refactor to use for loop or helper function

I have made another helper function